File: 2022/2_Rock_Paper_Scissors/part_1.ts
import { readFileSync } from 'fs'
import { map, reduce, split } from 'lodash'
import { Game, OutcomeValueMapping } from './types'

const outcomeValueMapping: OutcomeValueMapping = {
  A: {
    X: 1 + 3, // Rock    + Draw
    Y: 2 + 6, // Paper   + Won
    Z: 3 + 0, // Scizors + Lost
  },
  B: {
    X: 1 + 0, // Rock    + Lost
    Y: 2 + 3, // Paper   + Draw
    Z: 3 + 6, // Scizors + Won
  },
  C: {
    X: 1 + 6, // Rock    + Won
    Y: 2 + 0, // Paper   + Lost
    Z: 3 + 3, // Scizors + Draw
  },
}

const games: Game[] = map(
  split(readFileSync('input.txt', 'utf-8'), '\n'),
  (game) => split(game, ' ') as Game
)

console.log(
  reduce(games, (sum, game) => sum + outcomeValueMapping[game[0]][game[1]], 0)
)
// => 9177

File: 2022/2_Rock_Paper_Scissors/part_1.rb
input =
  File
    .open('input.txt')
    .read
    .split("\n")
    .map { |interaction| interaction.split(' ') }

# --------------------
directory = {
  'A' => {
    'X' => 3 + 1,
    'Y' => 6 + 2,
    'Z' => 3,
  },
  'B' => {
    'X' => 1,
    'Y' => 3 + 2,
    'Z' => 6 + 3,
  },
  'C' => {
    'X' => 6 + 1,
    'Y' => 2,
    'Z' => 3 + 3,
  },
}

print input
        .map { |interaction| directory[interaction.first][interaction.last] }
        .sum
